### PR TITLE
Add support for audioPassThruIcon parameter

### PR DESCRIPTION
**Pull request by tkireva**
New Image type parameter is added to MOBILE_API and HMI_API.
It is processed on audioPassThru request from the mobile side
and if it is valid, it is transmited to the HMI side.
Resolution of UI and TTS result codes received from the HMI
and their transfer to the mobile side is changed due to the
new CRQ.

Related-issues: [APPLINK-23360](https://adc.luxoft.com/jira/browse/APPLINK-23360)
Implements: [APPLINK-23358](https://adc.luxoft.com/jira/browse/APPLINK-23358)

### CODE REVIEW

**Comment by tkireva** [comment]
Please review:
@vlantonov @dtrunov @LuxoftAKutsan @TMelnyk

File: src/components/application_manager/include/application_manager/commands/mobile/perform_audio_pass_thru_request.h [hunk 1]
```diff
@@ -133,11 +133,65 @@ class PerformAudioPassThruRequest : public CommandRequestImpl {
    */
   bool IsWaitingHMIResponse();
 
+  /**
+   * @brief Validates audioPassThruIcon parameter and
+   * removes it if is not valid
+   * @param Pointer to the application whose storage directory
```
Comment (dtrunov) [inline src/components/application_manager/include/application_manager/commands/mobile/perform_audio_pass_thru_request.h]: @tkireva omitted name of parameter

File: src/components/application_manager/include/application_manager/commands/mobile/perform_audio_pass_thru_request.h [hunk 2]
```diff
@@ -133,11 +133,65 @@ class PerformAudioPassThruRequest : public CommandRequestImpl {
...
+   * @return If the audioPassthruIcon parameter value contains
+   * above mentioned symbols, the function returns FALSE, else
+   * the parameter value is valid and it returns TRUE
+   */
+  bool IsAudioPassThruIconParamValid();
+
+  /**
+   * @brief Maintains the full path filename for the
+   * audioPassThruIcon file
+   * @param Pointer to the application whose storage directory
```
Comment (dtrunov) [inline src/components/application_manager/include/application_manager/commands/mobile/perform_audio_pass_thru_request.h]: @tkireva Read comment above

File: src/components/application_manager/include/application_manager/commands/mobile/perform_audio_pass_thru_request.h [hunk 3]
```diff
@@ -133,11 +133,65 @@ class PerformAudioPassThruRequest : public CommandRequestImpl {
...
+   * @return result code - 1) UI error code has precedence than TTS's
+   * 2) error_code from TTS is turned to WARNINGS
+   */
+  mobile_apis::Result::eType PrepareAudioPassThruResultCodeForResponse(
+      const ResponseInfo& ui_response, const ResponseInfo& tts_response);
+
+  /**
+   * @brief Checks if any of audioPassThru communication components
+   * returned ABORTED code
+   * @param ui_response contains result_code from UI
```
Comment (dtrunov) [inline src/components/application_manager/include/application_manager/commands/mobile/perform_audio_pass_thru_request.h]: @tkireva params have incorrect name in description

File: src/components/application_manager/src/commands/mobile/perform_audio_pass_thru_request.cc [hunk 4]
```diff
@@ -365,6 +387,93 @@ bool PerformAudioPassThruRequest::IsWaitingHMIResponse() {
...
 
+void PerformAudioPassThruRequest::ProcessAudioPassThruIcon(
+    ApplicationSharedPtr app) {
+  LOG4CXX_AUTO_TRACE(logger_);
+  audio_pass_thru_icon_exists_ = true;
+  if ((*message_)[strings::msg_params].keyExists(
+          strings::audio_pass_thru_icon)) {
+    if (IsAudioPassThruIconParamValid()) {
+      const std::string& icon_storage_file = GetAudioPassThruIconFilename(app);
+
```
Comment (dtrunov) [inline src/components/application_manager/src/commands/mobile/perform_audio_pass_thru_request.cc]: @tkireva Please investigate method MessageHelper::VerifyImage for verify image

File: src/components/application_manager/src/commands/mobile/perform_audio_pass_thru_request.cc [hunk 5]
```diff
@@ -193,21 +196,35 @@ bool PerformAudioPassThruRequest::PrepareResponseParameters(
...
   LOG4CXX_AUTO_TRACE(logger_);
 
+  bool result = true;
   ResponseInfo ui_perform_info(result_ui_, HmiInterfaces::HMI_INTERFACE_UI);
   ResponseInfo tts_perform_info(result_tts_speak_,
                                 HmiInterfaces::HMI_INTERFACE_TTS);
-  const bool result =
-      PrepareResultForMobileResponse(ui_perform_info, tts_perform_info);
+
+  PrepareResultForMobileResponse(ui_perform_info, tts_perform_info);
```
Comment (dtrunov) [inline src/components/application_manager/src/commands/mobile/perform_audio_pass_thru_request.cc]: @tkireva PrepareResultForMobileResponse calculate result for sending with response to mobile side. Why did you remove assignment to result?

**Comment by TMelnyk** [comment]
reviewed

File: src/components/application_manager/src/commands/mobile/perform_audio_pass_thru_request.cc [hunk 6]
```diff
@@ -365,6 +387,93 @@ bool PerformAudioPassThruRequest::IsWaitingHMIResponse() {
...
+  hmi_apis::Common_Result::eType common_result =
+      hmi_apis::Common_Result::INVALID_ENUM;
+  const hmi_apis::Common_Result::eType ui_result = ui_response.result_code;
+  const hmi_apis::Common_Result::eType tts_result = tts_response.result_code;
+
+  if ((ui_result == hmi_apis::Common_Result::SUCCESS) &&
+      (tts_result != hmi_apis::Common_Result::SUCCESS)) {
+    common_result = hmi_apis::Common_Result::WARNINGS;
+  } else if (tts_result == hmi_apis::Common_Result::ABORTED) {
+    common_result = hmi_apis::Common_Result::ABORTED;
```
Comment (dtrunov) [inline src/components/application_manager/src/commands/mobile/perform_audio_pass_thru_request.cc]: @tkireva I think this checking is redundant because above in method PrepareResponseParameters you have checking 
IsAnyHMIComponentAborted

File: src/components/application_manager/src/commands/mobile/perform_audio_pass_thru_request.cc (hunk 5 above)
Comment (tkireva) [inline src/components/application_manager/src/commands/mobile/perform_audio_pass_thru_request.cc]: @dtrunov 
As the new CRQ defines new result code resolution (in the common case UI result code is with higher priority and must be transfered to the mobile side), the success result that is returned from PrepareResultForMobileResponse is not the right one in many cases and it must be changed.

File: src/components/application_manager/src/commands/mobile/perform_audio_pass_thru_request.cc (hunk 6 above)
Comment (tkireva) [inline src/components/application_manager/src/commands/mobile/perform_audio_pass_thru_request.cc]: @dtrunov 
I agree, you are absolutely right

File: src/components/application_manager/src/commands/mobile/perform_audio_pass_thru_request.cc (hunk 4 above)
Comment (tkireva) [inline src/components/application_manager/src/commands/mobile/perform_audio_pass_thru_request.cc]: @dtrunov I can use this function to check if the image exists in the file system. String value parameter requires checking for special characters. Please, look at the changes.

File: src/components/application_manager/include/application_manager/commands/mobile/perform_audio_pass_thru_request.h (hunk 3 above)
Comment (tkireva) [inline src/components/application_manager/include/application_manager/commands/mobile/perform_audio_pass_thru_request.h]: @dtrunov done

File: src/components/application_manager/include/application_manager/commands/mobile/perform_audio_pass_thru_request.h (hunk 2 above)
Comment (tkireva) [inline src/components/application_manager/include/application_manager/commands/mobile/perform_audio_pass_thru_request.h]: @dtrunov done

File: src/components/application_manager/include/application_manager/commands/mobile/perform_audio_pass_thru_request.h (hunk 1 above)
Comment (tkireva) [inline src/components/application_manager/include/application_manager/commands/mobile/perform_audio_pass_thru_request.h]: @dtrunov done

**Comment by tkireva** [comment]
Rebuild required

File: src/components/application_manager/include/application_manager/commands/command_request_impl.h [hunk 7]
```diff
@@ -259,6 +259,20 @@ class CommandRequestImpl : public CommandImpl,
...
+   * UNSUPPORTED_RESOURCE
+   * @param first contains result_code from HMI response and
+   * interface that returns response
+   * @param second contains result_code from HMI response and
+   * interface that returns response.
+   * @return if the communication return code must be
+   * UNSUPPORTED_RESOURCE, function returns true, else
+   * false
+   */
+  bool IsResultCodeUnsupported(const ResponseInfo& first,
```
Comment (LuxoftAKutsan) [inline src/components/application_manager/include/application_manager/commands/command_request_impl.h]: @tkireva Do I understand correctly that first is UI response and second is TTS response? If yes, please give appropriate names. Note that we have more then 2 HMI interfaces.

File: src/components/application_manager/src/commands/mobile/perform_audio_pass_thru_request.cc (hunk 5 above)
Comment (LuxoftAKutsan) [inline src/components/application_manager/src/commands/mobile/perform_audio_pass_thru_request.cc]: @tkireva any static analyzer will warn that you ignore result of boolean function, Please read its result and use mack UNUSED for it

File: src/components/application_manager/src/commands/mobile/perform_audio_pass_thru_request.cc [hunk 8]
```diff
@@ -193,21 +196,35 @@ bool PerformAudioPassThruRequest::PrepareResponseParameters(
...
+    result_code = PrepareAudioPassThruResultCodeForResponse(ui_perform_info,
+                                                            tts_perform_info);
+    if (!ui_perform_info.is_ok) {
+      result = false;
+    }
   }
-  result_code = PrepareResultCodeForResponse(ui_perform_info, tts_perform_info);
+
   info = MergeInfos(ui_perform_info, ui_info_, tts_perform_info, tts_info_);
+  if (audio_pass_thru_icon_exists_ == false) {
```
Comment (LuxoftAKutsan) [inline src/components/application_manager/src/commands/mobile/perform_audio_pass_thru_request.cc]: @tkireva `!audio_pass_thru_icon_exists_`

File: src/components/application_manager/src/commands/mobile/perform_audio_pass_thru_request.cc [hunk 9]
```diff
@@ -193,21 +196,35 @@ bool PerformAudioPassThruRequest::PrepareResponseParameters(
...
+    info = MergeInfos("Reference image(s) not found", info);
```
Comment (LuxoftAKutsan) [inline src/components/application_manager/src/commands/mobile/perform_audio_pass_thru_request.cc]: @tkireva  do not rewrite already assigned variable. Use if else

File: src/components/application_manager/src/commands/mobile/perform_audio_pass_thru_request.cc [hunk 10]
```diff
@@ -273,6 +290,11 @@ void PerformAudioPassThruRequest::SendPerformAudioPassThruRequest() {
     msg_params[str::mute_audio] = true;
   }
 
+  if ((*message_)[str::msg_params].keyExists(str::audio_pass_thru_icon)) {
```
Comment (LuxoftAKutsan) [inline src/components/application_manager/src/commands/mobile/perform_audio_pass_thru_request.cc]: @tkireva why you don't use msg_params variable?

File: src/components/application_manager/src/commands/mobile/perform_audio_pass_thru_request.cc [hunk 11]
```diff
@@ -365,6 +387,77 @@ bool PerformAudioPassThruRequest::IsWaitingHMIResponse() {
   return awaiting_tts_speak_response_ || awaiting_ui_response_;
 }
 
+void PerformAudioPassThruRequest::ProcessAudioPassThruIcon(
```
Comment (LuxoftAKutsan) [inline src/components/application_manager/src/commands/mobile/perform_audio_pass_thru_request.cc]: @tkireva  adding to function 
```
using smart_objects::SmartObject;
msg_params = (*message_)[strings::msg_params];

```
will make it more readble

File: src/components/application_manager/include/application_manager/commands/command_request_impl.h (hunk 7 above)
Comment (tkireva) [inline src/components/application_manager/include/application_manager/commands/command_request_impl.h]: This function is common for all commands and the info parameters may be from all available HMI interfaces (not every time for UI and TTS). That is the reason why I didn't change the names of parameters. I have just added it to the header file.

File: src/components/application_manager/src/commands/mobile/perform_audio_pass_thru_request.cc (hunk 9 above)
Comment (tkireva) [inline src/components/application_manager/src/commands/mobile/perform_audio_pass_thru_request.cc]: The info variable in this case, using MergeInfos method, will be extended, not rewritten.

**Comment by tkireva** [comment]
@LuxoftAKutsan Requested changes are done in f2a12dfca6c08240ebffd865635844a78d78ab6c

**Comment by tkireva** [comment]
@LuxoftAKutsan requested changes according to the mentioned above ignored result from method PrepareResultForMobileResponse() are done in 045d16ca20999b8ed1828daed2100fa25da34856

**Comment by tkireva** [comment]
Rebuild required
